In [1]:
import pandas as pd
import numpy as np
import warnings
import seaborn as sns
import matplotlib.pyplot as plt
from math import sqrt

In [2]:
#incase of unnecessary errors
warnings.filterwarnings("ignore", category=pd.core.common.SettingWithCopyWarning)

#Import data
coffee_store=pd.DataFrame(pd.read_csv('C:/Program Files/PostgreSQL/15/data/data_copy/coffee_csv1.csv'))

In [3]:
coffee_store.head()

region             place_name   place_type  rating  reviews price  \
0      C               Dim Kavu       Others     4.6    206.0    $$   
1      C                Коферум         Cafe     5.0     24.0    $$   
2      C       Кофейня Світ Чаю  Coffee shop     5.0     11.0    $$   
3      C       Кофейня Starcoff  Coffee shop     4.4    331.0    $$   
4      C  Кофейня "Friend Zone"  Coffee shop     5.0     12.0    $$   

   delivery_option dine_in_option takeout_option  
0            False            NaN            NaN  
1            False            NaN           True  
2            False            NaN           True  
3            False           True           True  
4            False           True           True

In [4]:
coffee_store.tail()

region                   place_name   place_type  rating  reviews price  \
195      H                       Artist       Others     4.6    963.0   $$$   
196      H                   Dream Cafe         Cafe     4.3    646.0    $$   
197      H  РУТА КАФЕ ПП КУЛІЧЕНКО К.І.         Cafe     4.6     16.0    $$   
198      H               Ob'yektna Kava  Coffee shop     4.7    599.0    $$   
199      H   О...МАРАТ КАФЕ ПП ПОЛІСТЕП         Cafe     4.3    341.0    $$   

     delivery_option dine_in_option takeout_option  
195            False           True           True  
196            False           True           True  
197            False           True           True  
198            False           True           True  
199            False           True            NaN

In [5]:
coffee_store.sample(10)

region               place_name   place_type  rating  reviews price  \
8        C       Кав'ярня My coffee  Coffee shop     4.8     31.0    $$   
34       J  Lviv Handmade Chocolate       Others     4.6   1376.0    $$   
2        C         Кофейня Світ Чаю  Coffee shop     5.0     11.0    $$   
45       F             coffee House  Coffee shop     4.7    279.0    $$   
111      E    Saturdays Coffee Цирк  Coffee shop     5.0     24.0    $$   
181      H                 Dim Kavy       Others     4.9     39.0    $$   
43       F    Prostir.coffee Таврик  Coffee shop     5.0     26.0    $$   
182      H                   Бамбук         Cafe     4.1    189.0    $$   
174      I             Kava U Shafi  Coffee shop     4.6    702.0    $$   
189      H              The sisters         Cafe     4.7    155.0    $$   

     delivery_option dine_in_option takeout_option  
8              False           True           True  
34             False           True           True  
2              False            NaN           True  
45             False           True           True  
111             True           True           True  
181            False            NaN            NaN  
43             False           True            NaN  
182            False           True           True  
174            False           True            NaN  
189            False           True           True

In [6]:
coffee_store.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   region           200 non-null    object 
 1   place_name       200 non-null    object 
 2   place_type       200 non-null    object 
 3   rating           198 non-null    float64
 4   reviews          198 non-null    float64
 5   price            200 non-null    object 
 6   delivery_option  200 non-null    bool   
 7   dine_in_option   140 non-null    object 
 8   takeout_option   144 non-null    object 
dtypes: bool(1), float64(2), object(6)
memory usage: 12.8+ KB


In [7]:
coffee_store.describe(include='all')

region place_name   place_type      rating       reviews price  \
count     200        200          200  198.000000    198.000000   200   
unique     10        187            4         NaN           NaN     3   
top         C   Dim Kavy  Coffee shop         NaN           NaN    $$   
freq       20          4           97         NaN           NaN   194   
mean      NaN        NaN          NaN    4.663636    622.494949   NaN   
std       NaN        NaN          NaN    0.221648   1400.898666   NaN   
min       NaN        NaN          NaN    3.900000      3.000000   NaN   
25%       NaN        NaN          NaN    4.600000     47.500000   NaN   
50%       NaN        NaN          NaN    4.700000    271.500000   NaN   
75%       NaN        NaN          NaN    4.800000    786.250000   NaN   
max       NaN        NaN          NaN    5.000000  17937.000000   NaN   

       delivery_option dine_in_option takeout_option  
count              200            140            144  
unique               2              1              1  
top              False           True           True  
freq               165            140            144  
mean               NaN            NaN            NaN  
std                NaN            NaN            NaN  
min                NaN            NaN            NaN  
25%                NaN            NaN            NaN  
50%                NaN            NaN            NaN  
75%                NaN            NaN            NaN  
max                NaN            NaN            NaN

In [8]:
#Count null
coffee_store.isnull().sum()

region              0
place_name          0
place_type          0
rating              2
reviews             2
price               0
delivery_option     0
dine_in_option     60
takeout_option     56
dtype: int64

In [9]:
coffee_store.drop_duplicates(inplace=True)
coffee_store.isnull().sum()

region              0
place_name          0
place_type          0
rating              2
reviews             2
price               0
delivery_option     0
dine_in_option     60
takeout_option     56
dtype: int64

In [10]:
coffee_store['dine_in_option'].fillna(False, inplace=True)
coffee_store['takeout_option'].fillna(False, inplace=True)

In [11]:
coffee_store.dropna(inplace=True)
coffee_store.isnull().sum()

region             0
place_name         0
place_type         0
rating             0
reviews            0
price              0
delivery_option    0
dine_in_option     0
takeout_option     0
dtype: int64

In [12]:
coffee_store['reviews'] = coffee_store['reviews'].astype(int)
coffee_store.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 198 entries, 0 to 199
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   region           198 non-null    object 
 1   place_name       198 non-null    object 
 2   place_type       198 non-null    object 
 3   rating           198 non-null    float64
 4   reviews          198 non-null    int32  
 5   price            198 non-null    object 
 6   delivery_option  198 non-null    bool   
 7   dine_in_option   198 non-null    bool   
 8   takeout_option   198 non-null    bool   
dtypes: bool(3), float64(1), int32(1), object(4)
memory usage: 10.6+ KB


#                                Rating analysis

In [13]:
value = coffee_store.groupby(['rating', 'place_type']).size().unstack(fill_value=0)
print(value)

place_type  Cafe  Coffee shop  Espresso bar  Others
rating                                             
3.9            1            0             0       1
4.0            2            0             0       0
4.1            1            1             0       0
4.2            2            0             0       0
4.3            3            3             0       0
4.4            5            7             0       0
4.5            9            8             3       2
4.6           10           24             7       6
4.7            9           21             2       7
4.8            2           11             5       0
4.9            6           13             3       3
5.0            7            8             0       6


In [14]:
# Find the average
mean_rating = np.mean(coffee_store['rating'])
mean_rating

4.663636363636367

In [15]:
# Find the median
np.median(coffee_store['rating'])

4.7

In [16]:
# Find the mode
coffee_store['rating'].mode()

0    4.6
dtype: float64

In [17]:
# Find the variance
np.var(coffee_store['rating'], ddof=1)

0.04912782648823264

In [18]:
# Find standard deviation
stddev=np.std(coffee_store['rating'], ddof=1)
stddev

0.22164797875963732

In [19]:
# Find the number of rating
sqrt_count=sqrt(len(coffee_store['rating']))
sqrt_count

14.071247279470288

In [20]:
# Find the standard error
round(stddev/sqrt_count,2)

0.02

In [21]:
# Find the upper and lower confidence interval
confidence_interval = {}
confidence_interval['confidence_lower'] = mean_rating-1.96 *stddev/sqrt_count
confidence_interval['confidence_upper'] = mean_rating+1.96 *stddev/sqrt_count
confidence_interval

{'confidence_lower': 4.632762765174152, 'confidence_upper': 4.694509962098581}

     OBSERVATION
The mean rating for the dataset is 4.66, and the calculated 95% confidence interval

for the mean is approximately [4.6327,4.6945]. This interval suggests that, with 

95% confidence, the true population mean rating is likely to fall within this 
range.

# Frequency

In [22]:
value = coffee_store.groupby(['delivery_option', 'place_type']).size().unstack(fill_value=0)
print(value)

place_type       Cafe  Coffee shop  Espresso bar  Others
delivery_option                                         
False              42           87            17      17
True               15            9             3       8


In [23]:
value = coffee_store.groupby(['dine_in_option', 'place_type']).size().unstack(fill_value=0)
print(value)

place_type      Cafe  Coffee shop  Espresso bar  Others
dine_in_option                                         
False             11           21            10      16
True              46           75            10       9


In [24]:
value = coffee_store.groupby(['takeout_option', 'place_type']).size().unstack(fill_value=0)
print(value)

place_type      Cafe  Coffee shop  Espresso bar  Others
takeout_option                                         
False              9           20             8      17
True              48           76            12       8


        OBSERVATION
1. The data suggests that the "Coffee shop" category tends to have a high 
frequency of  establishments offering both dine-in and take-out options.
2. "Cafe" shows a higher frequency in offering delivery.
3. The "Others" category seems to have varied offerings in terms of delivery, 
dine-in, and take-out.


# Review Central Tendency

In [25]:
value = coffee_store.groupby(['region'])[['reviews','rating']].agg([min,max])

print(value)

       reviews        rating     
           min    max    min  max
region                           
A           10   2700    4.5  5.0
B           11   1656    4.2  5.0
C           10    806    4.2  5.0
D           33   1509    4.0  5.0
E           10  17937    4.3  5.0
F           15   1345    4.4  5.0
G           21   2319    4.6  5.0
H            9   2381    3.9  5.0
I            3   2914    3.9  4.8
J           12   1785    4.1  5.0


     OBSERVATIONS 

1. The region with the highest maximum reviews is E, achieving a perfect rating
of 5. In contrast, region C has the lowest maximum reviews (806), but it also 
maintains a perfect rating of 5. Notably, region I is the only one with a maximum
rating lower than 5.

2. Recognizing the challenge of distinguishing between good and bad reviews, 
less emphasis is placed on reviews in the analysis.

3. Across regions, there is a consistent trend of a maximum rating of 5. To 
capture subtle variations, the minimum rating for each region is considered, as 
it offers greater accuracy.

4. Region G stands out with the highest minimum rating of 4.6, while both 
regions I and H share the lowest minimum rating.
